.drone.yml

```diff
@@ -0,0 +1,1100 @@
+---
+kind: pipeline
+type: kubernetes
+name: test
+
+trigger:
+  branch:
+    - master
+    - branch/*
+  event:
+    exclude:
+      - promote
+      - rollback
+
+workspace:
+  path: /go
+
+clone:
+  disable: true
+
+steps:
+  - name: Check out code
+    image: golang:1.13.2
+    environment:
+      GITHUB_PRIVATE_KEY:
+        from_secret: GITHUB_PRIVATE_KEY
+    commands:
+      - mkdir -p /go/src/github.com/gravitational/teleport
+      - cd /go/src/github.com/gravitational/teleport
+      - git clone https://github.com/gravitational/teleport.git .
+      - git checkout $DRONE_COMMIT
+      - echo $DRONE_SOURCE_BRANCH > /go/.drone_source_branch.txt
+      # fetch enterprise submodules
+      - mkdir -m 0700 /root/.ssh && echo -n "$GITHUB_PRIVATE_KEY" > /root/.ssh/id_rsa && chmod 600 /root/.ssh/id_rsa
+      - ssh-keyscan -H github.com > /root/.ssh/known_hosts 2>/dev/null && chmod 600 /root/.ssh/known_hosts
+      - git submodule update --init e
+      - git submodule update --init --recursive webassets || true
+      - rm -f /root/.ssh/id_rsa
+
+  - name: Restore cache
+    image: meltwater/drone-cache
```

Comment:
I'm probably paranoid, but this is another supply chain risk.
This container will have access to our AWS credentials. Do these only have RW access to one bucket?

Reply:
Yes - the AWS credentials which are available to pull requests only have access to the dronestorage “staging” bucket and nothing else. They’re limited by an IAM role.

The production credentials aren’t available to pull requests and can only be accessed by an authorised user running a promotion command from within Drone itself.

Reply:
(pull requests will also be unsigned and so have to be approved by an admin before they run - we can look at the code and decline if anyone is trying to dump credentials to `stdout` or something)

Reply:
I'm mostly concerned about someone compromising the `meltwater` dockerhub account.
The attack scenario here is pretty sophisticated:
- pwn dockerhub account
- upload a malicious `drone-cache` image
- this image exfiltrates our staging s3 credentials
- attacker uploads a backdored version of build artifacts to staging s3 bucket
- authorized user promotes this to production during release
- our users pull down promoted binaries and get pwned through a backdoor

The attacker would need to target teleport specifically. It also assumes we don't catch anything during release testing.

Since all of that is very unlikely, we can probably keep it as is.
Maybe in some distant future we can revisit our CI dependencies.

Reply:
That’s a valid point. It is a very sophisticated and unlikely set of actions as you say, but possible nonetheless.

The code is public (https://github.com/meltwater/drone-cache) and licensed under Apache 2.0, so we could fork the repo, build the plugin and upload our own image to prevent this in future if we want.

Reply:
I'm not sure if this is a uncommon attack vector, haven't popular NPM modules been hit by a similar attack multiple times recently? I believe in those cases the authors gave up ownership and moved on to other things, but the underlying principal is the same: we should be very careful about from whom we accept external dependencies.

I think we can take two approaches:

1. We remove the build cache. I think this is a feasible alternative because cutting builds is not a frequent operation and even if we double or triple the time to cut a build, it won't slow our velocity down. We spend substantially more time on design and development than cutting builds.

2. We maintain our own fork of the image.

What do you guys think?

Reply:
To be clear - this is a cache for tests rather than builds. We always build artifacts clean from source and don't use any caching.

**Pros:**
- The cache does make iterative testing substantially faster; it cuts the time for feedback roughly in half - about 7-8 minutes per commit with caching versus maybe 15 minutes or so without caching.

**Cons:**
- Hypothetical attack vector as described (which could be mitigated by us self-building/hosting the image)
- We have to run the linter/unit tests/integration tests sequentially to be able to make use of caching, whereas if we disabled caching we could run all of these 3 in parallel and the time taken would be about equivalent to the longest running tests (which I think are currently integration tests at ~6/7 minutes all told, I haven't checked in a while though)
- Caching has diminishing returns as PRs get up into the hundreds of commits - for example, the cache for this PR is now up to about 650MB compressed and it's taking ~30 seconds to download/decompress the cache and ~3 minutes to recompress/upload it again at the end. However, it's very rare for us to have PRs with ~200 commits and the improvements in feedback time lower down are good for quick development and confidence.

Maintaining our own fork of the image would not be prohibitively difficult if we wanted to go that way. I'm happy to disable caching entirely if we think that's a better solution. The caching would also make tests on the master/release branches substantially faster as we'd effectively only be recompiling the deltas between each push rather than the entirely of Teleport each time.

Reply:
It seems like the caching is actually doing more harm than good on this branch.

Caching on:
<img width="280" alt="Screenshot 2020-06-24 17 29 18" src="https://user-images.githubusercontent.com/897821/85625903-91e4a900-b642-11ea-84fa-0d92f768f30b.png">

Caching off:
<img width="284" alt="Screenshot 2020-06-24 17 29 10" src="https://user-images.githubusercontent.com/897821/85625943-a163f200-b642-11ea-88c5-0038f9bdd1d3.png">

Reply:
I'll just remove all the caching and then we don't have to worry.